Catalog Downloading state getting stuck

**Describe the bug**
Catalog - installing Addon with missing stable channel causing it to halt on "downloading" state

15:48:23.294 [ajour::gui::update][DEBUG] Message::ParsedAddons(classic, 52 addons)
15:48:28.732 [ajour::gui::update][DEBUG] Interaction::ModeSelected(Catalog)
15:49:37.538 [ajour::gui::update][DEBUG] Interaction::CatalogInstall(Curse, classic, 399223)
15:49:37.657 [ajour::gui::update][ERROR] No stable file found for curse id 399223

The addon i tried to install.
https://www.curseforge.com/wow/addons/softres/files

**To Reproduce**
Catalog -> Search "Softres" > Install the one with about 2k Downloads by clicking "Classic"

**Expected behavior**
Should leave downloading state

**Screenshots**
![image](https://user-images.githubusercontent.com/17293360/95472141-f5d97600-0982-11eb-84f0-5236ac6f0a0e.png)

**Software involved**
Windows 10 - 2004
Ajour 0.4.0 Release

**Log Output**
[ajour.log](https://github.com/casperstorm/ajour/files/5348562/ajour.log)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ The sections should follow the order `Packaging`, `Added`, `Changed`, `Fixed` an
 
 ### Fixed
 - Fixed bug where orphaned folders could exist after updating an addon if the newer version of an addon didnt't include those folders anymore. 
+- Fixed catalog install buttons getting stuck when install fails or addon is unavailable to download. Button will now show "Retry" if failed and disabled as "Unavailable" if the addon is unavailable.
 
 ### Packaging
 - Added Forest Night theme

--- a/crates/core/src/curse_api.rs
+++ b/crates/core/src/curse_api.rs
@@ -163,7 +163,7 @@ pub async fn fetch_game_info() -> Result<GameInfo> {
     }
 }
 
-pub async fn latest_addon(curse_id: u32, flavor: Flavor) -> Result<(u32, Flavor, Addon)> {
+pub async fn latest_addon(curse_id: u32, flavor: Flavor) -> Result<Addon> {
     let packages: Vec<Package> = fetch_remote_packages_by_ids(&[curse_id]).await?;
 
     let package = packages.into_iter().next().ok_or_else(|| {
@@ -210,5 +210,5 @@ pub async fn latest_addon(curse_id: u32, flavor: Flavor) -> Result<(u32, Flavor,
 
     addon.set_title(package.name.clone());
 
-    Ok((curse_id, flavor, addon))
+    Ok(addon)
 }

--- a/crates/core/src/parse.rs
+++ b/crates/core/src/parse.rs
@@ -498,7 +498,6 @@ pub async fn read_addon_directory<P: AsRef<Path>>(
         .cloned();
 
     let unknown_addons = unmapped_folders
-        .into_iter()
         .map(|f| {
             let mut addon = Addon::empty(&f.id);
             addon.folders = vec![f];

--- a/crates/core/src/tukui_api.rs
+++ b/crates/core/src/tukui_api.rs
@@ -72,14 +72,14 @@ pub async fn fetch_remote_package(id: &str, flavor: &Flavor) -> Result<TukuiPack
     }
 }
 
-pub async fn latest_addon(tukui_id: u32, flavor: Flavor) -> Result<(u32, Flavor, Addon)> {
+pub async fn latest_addon(tukui_id: u32, flavor: Flavor) -> Result<Addon> {
     let tukui_id_string = tukui_id.to_string();
 
     let package = fetch_remote_package(&tukui_id_string, &flavor).await?;
 
     let addon = Addon::from_tukui_package(tukui_id_string, &[], &package);
 
-    Ok((tukui_id, flavor, addon))
+    Ok(addon)
 }
 
 pub async fn fetch_changelog(id: &str, flavor: &Flavor) -> Result<(String, String)> {

--- a/src/gui/element.rs
+++ b/src/gui/element.rs
@@ -3,9 +3,9 @@
 use {
     super::{
         style, AddonVersionKey, AjourMode, AjourState, BackupState, CatalogColumnKey,
-        CatalogColumnState, CatalogRow, Changelog, ColumnKey, ColumnSettings, ColumnState,
-        DirectoryType, ExpandType, Interaction, Message, ReleaseChannel, ScaleState, SortDirection,
-        ThemeState,
+        CatalogColumnState, CatalogInstallStatus, CatalogRow, Changelog, ColumnKey, ColumnSettings,
+        ColumnState, DirectoryType, ExpandType, Interaction, Message, ReleaseChannel, ScaleState,
+        SortDirection, ThemeState,
     },
     crate::VERSION,
     ajour_core::{
@@ -1453,10 +1453,9 @@ pub fn catalog_data_cell<'a, 'b>(
     color_palette: ColorPalette,
     addon: &'a mut CatalogRow,
     column_config: &'b [(CatalogColumnKey, Length)],
-    retail_downloading: bool,
     retail_installed: bool,
-    classic_downloading: bool,
     classic_installed: bool,
+    statuses: Vec<(Flavor, CatalogInstallStatus)>,
 ) -> Container<'a, Message> {
     let default_height = Length::Units(26);
 
@@ -1482,14 +1481,29 @@ pub fn catalog_data_cell<'a, 'b>(
         })
         .next()
     {
+        let status = statuses
+            .iter()
+            .find(|(f, _)| *f == Flavor::Retail)
+            .map(|(_, status)| *status);
+
         let retail_install = Text::new(if !retail_exists {
             "N/A"
-        } else if retail_downloading {
-            "Downloading"
-        } else if retail_installed {
-            "Installed"
         } else {
-            "Retail"
+            match status {
+                Some(CatalogInstallStatus::Downloading) => "Downloading",
+                Some(CatalogInstallStatus::Unpacking) => "Unpacking",
+                Some(CatalogInstallStatus::Fingerprint) => "Hashing",
+                Some(CatalogInstallStatus::Completed) => "Installed",
+                Some(CatalogInstallStatus::Retry) => "Retry",
+                Some(CatalogInstallStatus::Unavilable) => "Unavailable",
+                None => {
+                    if retail_installed {
+                        "Installed"
+                    } else {
+                        "Retail"
+                    }
+                }
+            }
         })
         .size(DEFAULT_FONT_SIZE);
 
@@ -1502,7 +1516,10 @@ pub fn catalog_data_cell<'a, 'b>(
             .style(style::DefaultButton(color_palette))
             .width(*width);
 
-        if !retail_installed && !retail_downloading && retail_exists {
+        if retail_exists
+            && (status == Some(CatalogInstallStatus::Retry)
+                || (status == None && !retail_installed))
+        {
             retail_install_button = retail_install_button.on_press(Interaction::CatalogInstall(
                 addon_data.source,
                 Flavor::Retail,
@@ -1534,14 +1551,29 @@ pub fn catalog_data_cell<'a, 'b>(
         })
         .next()
     {
+        let status = statuses
+            .iter()
+            .find(|(f, _)| *f == Flavor::Classic)
+            .map(|(_, status)| *status);
+
         let classic_install = Text::new(if !classic_exists {
             "N/A"
-        } else if classic_downloading {
-            "Downloading"
-        } else if classic_installed {
-            "Installed"
         } else {
-            "Classic"
+            match status {
+                Some(CatalogInstallStatus::Downloading) => "Downloading",
+                Some(CatalogInstallStatus::Unpacking) => "Unpacking",
+                Some(CatalogInstallStatus::Fingerprint) => "Hashing",
+                Some(CatalogInstallStatus::Completed) => "Installed",
+                Some(CatalogInstallStatus::Retry) => "Retry",
+                Some(CatalogInstallStatus::Unavilable) => "Unavailable",
+                None => {
+                    if classic_installed {
+                        "Installed"
+                    } else {
+                        "Classic"
+                    }
+                }
+            }
         })
         .size(DEFAULT_FONT_SIZE);
 
@@ -1555,7 +1587,10 @@ pub fn catalog_data_cell<'a, 'b>(
                 .style(style::DefaultButton(color_palette))
                 .width(*width);
 
-        if !classic_installed && !classic_downloading && classic_exists {
+        if classic_exists
+            && (status == Some(CatalogInstallStatus::Retry)
+                || (status == None && !classic_installed))
+        {
             classic_install_button = classic_install_button.on_press(Interaction::CatalogInstall(
                 addon_data.source,
                 Flavor::Classic,

--- a/src/gui/mod.rs
+++ b/src/gui/mod.rs
@@ -103,7 +103,7 @@ pub enum Message {
     None(()),
     Parse(Result<Config>),
     ParsedAddons((Flavor, Result<Vec<Addon>>)),
-    UpdateFingerprint((Flavor, String, Result<()>)),
+    UpdateFingerprint((DownloadReason, Flavor, String, Result<()>)),
     ThemeSelected(String),
     ReleaseChannelSelected(ReleaseChannel),
     ThemesLoaded(Vec<Theme>),
@@ -1122,7 +1122,9 @@ impl Default for CatalogSearchState {
 pub struct CatalogRow {
     website_state: button::State,
     retail_install_state: button::State,
+    retail_install_status: CatalogInstallStatus,
     classic_install_state: button::State,
+    classic_install_status: CatalogInstallStatus,
     addon: CatalogAddon,
 }
 
@@ -1131,10 +1133,23 @@ impl From<CatalogAddon> for CatalogRow {
         Self {
             website_state: Default::default(),
             retail_install_state: Default::default(),
+            retail_install_status: CatalogInstallStatus::None,
             classic_install_state: Default::default(),
+            classic_install_status: CatalogInstallStatus::None,
             addon,
         }
     }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum CatalogInstallStatus {
+    None,
+    Downloading,
+    Unpacking,
+    Fingerprint,
+    Completed,
+    Retry,
+    Unavilable,
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]

--- a/src/gui/mod.rs
+++ b/src/gui/mod.rs
@@ -114,7 +114,7 @@ pub enum Message {
     LatestBackup(Option<NaiveDateTime>),
     BackupFinished(Result<NaiveDateTime>),
     CatalogDownloaded(Result<Catalog>),
-    CatalogInstallAddonFetched(Result<(u32, Flavor, Addon)>),
+    CatalogInstallAddonFetched((Flavor, u32, Result<Addon>)),
     FetchedCurseChangelog((Addon, AddonVersionKey, Result<(String, String)>)),
     FetchedTukuiChangelog((Addon, AddonVersionKey, Result<(String, String)>)),
 }

--- a/src/gui/mod.rs
+++ b/src/gui/mod.rs
@@ -5,7 +5,7 @@ mod update;
 use crate::cli::Opts;
 use crate::VERSION;
 use ajour_core::{
-    addon::{Addon, AddonFolder, AddonState, AddonVersionKey, ReleaseChannel},
+    addon::{Addon, AddonFolder, AddonVersionKey, ReleaseChannel},
     catalog::get_catalog,
     catalog::{self, Catalog, CatalogAddon},
     config::{load_config, ColumnConfigV2, Config, Flavor},
@@ -146,6 +146,7 @@ pub struct Ajour {
     column_settings: ColumnSettings,
     onboarding_directory_btn_state: button::State,
     catalog: Option<Catalog>,
+    catalog_install_statuses: Vec<(Flavor, u32, CatalogInstallStatus)>,
     catalog_search_state: CatalogSearchState,
     catalog_header_state: CatalogHeaderState,
 }
@@ -185,6 +186,7 @@ impl Default for Ajour {
             column_settings: Default::default(),
             onboarding_directory_btn_state: Default::default(),
             catalog: None,
+            catalog_install_statuses: vec![],
             catalog_search_state: Default::default(),
             catalog_header_state: Default::default(),
         }
@@ -523,19 +525,6 @@ impl Application for Ajour {
                                     || a.tukui_id() == Some(&addon.addon.id.to_string())
                             })
                         };
-                        let retail_downloading = if flavor == Flavor::Retail {
-                            addons.iter().any(|a| {
-                                (a.curse_id() == Some(addon.addon.id)
-                                    || a.tukui_id() == Some(&addon.addon.id.to_string()))
-                                    && a.state == AddonState::Downloading
-                            })
-                        } else {
-                            other_flavor_addons.iter().any(|a| {
-                                (a.curse_id() == Some(addon.addon.id)
-                                    || a.tukui_id() == Some(&addon.addon.id.to_string()))
-                                    && a.state == AddonState::Downloading
-                            })
-                        };
 
                         let classic_installed = if flavor == Flavor::Classic {
                             addons.iter().any(|a| {
@@ -548,28 +537,21 @@ impl Application for Ajour {
                                     || a.tukui_id() == Some(&addon.addon.id.to_string())
                             })
                         };
-                        let classic_downloading = if flavor == Flavor::Classic {
-                            addons.iter().any(|a| {
-                                (a.curse_id() == Some(addon.addon.id)
-                                    || a.tukui_id() == Some(&addon.addon.id.to_string()))
-                                    && a.state == AddonState::Downloading
-                            })
-                        } else {
-                            other_flavor_addons.iter().any(|a| {
-                                (a.curse_id() == Some(addon.addon.id)
-                                    || a.tukui_id() == Some(&addon.addon.id.to_string()))
-                                    && a.state == AddonState::Downloading
-                            })
-                        };
+
+                        let statuses = self
+                            .catalog_install_statuses
+                            .iter()
+                            .filter(|(_, i, _)| addon.addon.id == *i)
+                            .map(|(flavor, _, status)| (*flavor, *status))
+                            .collect();
 
                         let catalog_data_cell = element::catalog_data_cell(
                             color_palette,
                             addon,
                             &catalog_column_config,
-                            retail_downloading,
                             retail_installed,
-                            classic_downloading,
                             classic_installed,
+                            statuses,
                         );
 
                         catalog_scrollable = catalog_scrollable.push(catalog_data_cell);
@@ -1122,9 +1104,7 @@ impl Default for CatalogSearchState {
 pub struct CatalogRow {
     website_state: button::State,
     retail_install_state: button::State,
-    retail_install_status: CatalogInstallStatus,
     classic_install_state: button::State,
-    classic_install_status: CatalogInstallStatus,
     addon: CatalogAddon,
 }
 
@@ -1133,17 +1113,14 @@ impl From<CatalogAddon> for CatalogRow {
         Self {
             website_state: Default::default(),
             retail_install_state: Default::default(),
-            retail_install_status: CatalogInstallStatus::None,
             classic_install_state: Default::default(),
-            classic_install_status: CatalogInstallStatus::None,
             addon,
         }
     }
 }
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Debug, Clone, Copy, PartialEq)]
 pub enum CatalogInstallStatus {
-    None,
     Downloading,
     Unpacking,
     Fingerprint,

--- a/src/gui/update.rs
+++ b/src/gui/update.rs
@@ -1,8 +1,9 @@
 use {
     super::{
         AddonVersionKey, Ajour, AjourMode, AjourState, CatalogCategory, CatalogColumnKey,
-        CatalogFlavor, CatalogRow, CatalogSource, Changelog, ChangelogPayload, ColumnKey,
-        DirectoryType, DownloadReason, ExpandType, Interaction, Message, SortDirection,
+        CatalogFlavor, CatalogInstallStatus, CatalogRow, CatalogSource, Changelog,
+        ChangelogPayload, ColumnKey, DirectoryType, DownloadReason, ExpandType, Interaction,
+        Message, SortDirection,
     },
     ajour_core::{
         addon::{Addon, AddonFolder, AddonState, Repository},
@@ -621,32 +622,88 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                 );
             }
         }
-        Message::DownloadedAddon((reason, flavor, id, Ok(()))) => {
-            log::debug!("Message::DownloadedAddon(({}, {}))", flavor, &id,);
+        Message::DownloadedAddon((reason, flavor, id, result)) => {
+            log::debug!(
+                "Message::DownloadedAddon(({}, {}, error: {}))",
+                flavor,
+                &id,
+                result.is_err()
+            );
 
-            // When an addon has been successfully downloaded we begin to unpack it.
-            // If it for some reason fails to download, we handle the error.
-            let from_directory = ajour
-                .config
-                .get_download_directory_for_flavor(flavor)
-                .expect("Expected a valid path");
-            let to_directory = ajour
-                .config
-                .get_addon_directory_for_flavor(&flavor)
-                .expect("Expected a valid path");
-            let addons = ajour.addons.entry(flavor).or_default();
-            if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
-                if addon.state == AddonState::Downloading {
-                    addon.state = AddonState::Unpacking;
-                    let addon = addon.clone();
-                    return Ok(Command::perform(
-                        perform_unpack_addon(reason, flavor, addon, from_directory, to_directory),
-                        Message::UnpackedAddon,
-                    ));
+            match result {
+                Ok(_) => {
+                    // When an addon has been successfully downloaded we begin to unpack it.
+                    // If it for some reason fails to download, we handle the error.
+                    let from_directory = ajour
+                        .config
+                        .get_download_directory_for_flavor(flavor)
+                        .expect("Expected a valid path");
+                    let to_directory = ajour
+                        .config
+                        .get_addon_directory_for_flavor(&flavor)
+                        .expect("Expected a valid path");
+
+                    if reason == DownloadReason::Install {
+                        if let Some(row) = ajour
+                            .catalog_search_state
+                            .catalog_rows
+                            .iter_mut()
+                            .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                        {
+                            match flavor {
+                                Flavor::Retail => {
+                                    row.retail_install_status = CatalogInstallStatus::Unpacking
+                                }
+                                Flavor::Classic => {
+                                    row.classic_install_status = CatalogInstallStatus::Unpacking
+                                }
+                            }
+                        }
+                    }
+
+                    let addons = ajour.addons.entry(flavor).or_default();
+                    if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
+                        if addon.state == AddonState::Downloading {
+                            addon.state = AddonState::Unpacking;
+                            let addon = addon.clone();
+                            return Ok(Command::perform(
+                                perform_unpack_addon(
+                                    reason,
+                                    flavor,
+                                    addon,
+                                    from_directory,
+                                    to_directory,
+                                ),
+                                Message::UnpackedAddon,
+                            ));
+                        }
+                    }
+                }
+                Err(error) => {
+                    log::error!("{}", error);
+                    ajour.state = AjourState::Error(error);
+
+                    if reason == DownloadReason::Install {
+                        if let Some(row) = ajour
+                            .catalog_search_state
+                            .catalog_rows
+                            .iter_mut()
+                            .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                        {
+                            match flavor {
+                                Flavor::Retail => {
+                                    row.retail_install_status = CatalogInstallStatus::Retry
+                                }
+                                Flavor::Classic => {
+                                    row.classic_install_status = CatalogInstallStatus::Retry
+                                }
+                            }
+                        }
+                    }
                 }
             }
         }
-        Message::UnpackedAddon((_reason, flavor, id, result)) => {
+        Message::UnpackedAddon((reason, flavor, id, result)) => {
             log::debug!(
                 "Message::UnpackedAddon(({}, error: {}))",
                 &id,
@@ -694,6 +751,26 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                             addon.folders = folders;
                         }
 
+                        if reason == DownloadReason::Install {
+                            if let Some(row) = ajour
+                                .catalog_search_state
+                                .catalog_rows
+                                .iter_mut()
+                                .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                            {
+                                match flavor {
+                                    Flavor::Retail => {
+                                        row.retail_install_status =
+                                            CatalogInstallStatus::Fingerprint
+                                    }
+                                    Flavor::Classic => {
+                                        row.classic_install_status =
+                                            CatalogInstallStatus::Fingerprint
+                                    }
+                                }
+                            }
+                        }
+
                         addon.state = AddonState::Fingerprint;
 
                         let mut version = None;
@@ -709,6 +786,7 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                         for folder in &addon.folders {
                             commands.push(Command::perform(
                                 perform_hash_addon(
+                                    reason,
                                     ajour
                                         .config
                                         .get_addon_directory_for_flavor(&flavor)
@@ -726,17 +804,53 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                     Err(err) => {
                         ajour.state = AjourState::Error(err);
                         addon.state = AddonState::Ajour(Some("Error".to_owned()));
+
+                        if reason == DownloadReason::Install {
+                            if let Some(row) = ajour
+                                .catalog_search_state
+                                .catalog_rows
+                                .iter_mut()
+                                .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                            {
+                                match flavor {
+                                    Flavor::Retail => {
+                                        row.retail_install_status = CatalogInstallStatus::Retry
+                                    }
+                                    Flavor::Classic => {
+                                        row.classic_install_status = CatalogInstallStatus::Retry
+                                    }
+                                }
+                            }
+                        }
                     }
                 }
             }
         }
-        Message::UpdateFingerprint((flavor, id, result)) => {
+        Message::UpdateFingerprint((reason, flavor, id, result)) => {
             log::debug!(
                 "Message::UpdateFingerprint(({:?}, {}, error: {}))",
                 flavor,
                 &id,
                 result.is_err()
             );
+
+            if reason == DownloadReason::Install {
+                if let Some(row) = ajour
+                    .catalog_search_state
+                    .catalog_rows
+                    .iter_mut()
+                    .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                {
+                    match flavor {
+                        Flavor::Retail => {
+                            row.retail_install_status = CatalogInstallStatus::Completed
+                        }
+                        Flavor::Classic => {
+                            row.classic_install_status = CatalogInstallStatus::Completed
+                        }
+                    }
+                }
+            }
 
             let addons = ajour.addons.entry(flavor).or_default();
             if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
@@ -1156,18 +1270,19 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             // Close settings if shown.
             ajour.is_showing_settings = false;
 
-            // We create an empty addon we can add to the list of addons.
-            // This will later be updated by a more rich addon.
-            let mut empty_addon = Addon::empty(&id.to_string());
-            empty_addon.state = AddonState::Downloading;
-
-            match source {
-                catalog::Source::Tukui => empty_addon.set_tukui_id(id.to_string()),
-                catalog::Source::Curse => empty_addon.set_curse_id(id),
+            if let Some(row) = ajour
+                .catalog_search_state
+                .catalog_rows
+                .iter_mut()
+                .find(|r| r.addon.id == id)
+            {
+                match flavor {
+                    Flavor::Retail => row.retail_install_status = CatalogInstallStatus::Downloading,
+                    Flavor::Classic => {
+                        row.classic_install_status = CatalogInstallStatus::Downloading
+                    }
+                }
             }
-
-            let addons = ajour.addons.entry(flavor).or_default();
-            addons.push(empty_addon);
 
             return Ok(Command::perform(
                 perform_fetch_latest_addon(source, id, flavor),
@@ -1223,14 +1338,7 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                 );
 
                 if let Some(addons) = ajour.addons.get_mut(&flavor) {
-                    // Remove the empty addon and add in our rich addon
-                    if addons
-                        .iter_mut()
-                        .any(|a| a.primary_folder_id == id.to_string())
-                    {
-                        addons.retain(|a| a.primary_folder_id != id.to_string());
-                    }
-
+                    // Add the addon to our collection
                     addon.state = AddonState::Downloading;
                     addons.push(addon.clone());
 
@@ -1254,9 +1362,20 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             Err(error) => {
                 log::error!("{}", error);
 
-                if let Some(addons) = ajour.addons.get_mut(&flavor) {
-                    // Remove the empty addon since it failed
-                    addons.retain(|a| a.primary_folder_id != id.to_string());
+                if let Some(row) = ajour
+                    .catalog_search_state
+                    .catalog_rows
+                    .iter_mut()
+                    .find(|r| r.addon.id == id)
+                {
+                    match flavor {
+                        Flavor::Retail => {
+                            row.retail_install_status = CatalogInstallStatus::Unavilable
+                        }
+                        Flavor::Classic => {
+                            row.classic_install_status = CatalogInstallStatus::Unavilable
+                        }
+                    }
                 }
             }
         },
@@ -1298,8 +1417,7 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
         Message::Error(error)
         | Message::Parse(Err(error))
         | Message::NeedsUpdate(Err(error))
-        | Message::CatalogDownloaded(Err(error))
-        | Message::DownloadedAddon((_, _, _, Err(error))) => {
+        | Message::CatalogDownloaded(Err(error)) => {
             log::error!("{}", error);
 
             ajour.state = AjourState::Error(error);
@@ -1381,12 +1499,14 @@ async fn perform_download_addon(
 
 /// Rehashes a `Addon`.
 async fn perform_hash_addon(
+    reason: DownloadReason,
     addon_dir: impl AsRef<Path>,
     addon_id: String,
     fingerprint_collection: Arc<Mutex<Option<FingerprintCollection>>>,
     flavor: Flavor,
-) -> (Flavor, String, Result<()>) {
+) -> (DownloadReason, Flavor, String, Result<()>) {
     (
+        reason,
         flavor,
         addon_id.clone(),
         update_addon_fingerprint(fingerprint_collection, flavor, addon_dir, addon_id).await,

--- a/src/gui/update.rs
+++ b/src/gui/update.rs
@@ -630,39 +630,34 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                 result.is_err()
             );
 
-            match result {
-                Ok(_) => {
-                    // When an addon has been successfully downloaded we begin to unpack it.
-                    // If it for some reason fails to download, we handle the error.
-                    let from_directory = ajour
-                        .config
-                        .get_download_directory_for_flavor(flavor)
-                        .expect("Expected a valid path");
-                    let to_directory = ajour
-                        .config
-                        .get_addon_directory_for_flavor(&flavor)
-                        .expect("Expected a valid path");
+            // When an addon has been successfully downloaded we begin to unpack it.
+            // If it for some reason fails to download, we handle the error.
+            let from_directory = ajour
+                .config
+                .get_download_directory_for_flavor(flavor)
+                .expect("Expected a valid path");
+            let to_directory = ajour
+                .config
+                .get_addon_directory_for_flavor(&flavor)
+                .expect("Expected a valid path");
 
-                    if reason == DownloadReason::Install {
-                        if let Some(row) = ajour
-                            .catalog_search_state
-                            .catalog_rows
-                            .iter_mut()
-                            .find(|r| r.addon.id == id.parse::<u32>().unwrap())
-                        {
-                            match flavor {
-                                Flavor::Retail => {
-                                    row.retail_install_status = CatalogInstallStatus::Unpacking
-                                }
-                                Flavor::Classic => {
-                                    row.classic_install_status = CatalogInstallStatus::Unpacking
-                                }
+            let mut remove_catalog_addon = None;
+
+            let addons = ajour.addons.entry(flavor).or_default();
+            if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
+                match result {
+                    Ok(_) => {
+                        // Update catalog status for addon
+                        if reason == DownloadReason::Install {
+                            if let Some((_, _, status)) =
+                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                    flavor == *f && addon.repository_id() == Some(i.to_string())
+                                })
+                            {
+                                *status = CatalogInstallStatus::Unpacking;
                             }
                         }
-                    }
 
-                    let addons = ajour.addons.entry(flavor).or_default();
-                    if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
                         if addon.state == AddonState::Downloading {
                             addon.state = AddonState::Unpacking;
                             let addon = addon.clone();
@@ -678,29 +673,30 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                             ));
                         }
                     }
-                }
-                Err(error) => {
-                    log::error!("{}", error);
-                    ajour.state = AjourState::Error(error);
+                    Err(error) => {
+                        log::error!("{}", error);
 
-                    if reason == DownloadReason::Install {
-                        if let Some(row) = ajour
-                            .catalog_search_state
-                            .catalog_rows
-                            .iter_mut()
-                            .find(|r| r.addon.id == id.parse::<u32>().unwrap())
-                        {
-                            match flavor {
-                                Flavor::Retail => {
-                                    row.retail_install_status = CatalogInstallStatus::Retry
-                                }
-                                Flavor::Classic => {
-                                    row.classic_install_status = CatalogInstallStatus::Retry
-                                }
+                        ajour.state = AjourState::Error(error);
+
+                        // Update catalog status for addon
+                        if reason == DownloadReason::Install {
+                            if let Some((_, _, status)) =
+                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                    flavor == *f && addon.repository_id() == Some(i.to_string())
+                                })
+                            {
+                                *status = CatalogInstallStatus::Retry;
+
+                                remove_catalog_addon = Some(addon.primary_folder_id.clone());
                             }
                         }
                     }
                 }
+            }
+
+            // Remove catalog installed addon from addons since it failed
+            if let Some(id) = remove_catalog_addon {
+                addons.retain(|a| a.primary_folder_id != id)
             }
         }
         Message::UnpackedAddon((reason, flavor, id, result)) => {
@@ -709,6 +705,8 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                 &id,
                 result.is_err()
             );
+
+            let mut remove_catalog_addon = None;
 
             let addons = ajour.addons.entry(flavor).or_default();
             if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
@@ -751,23 +749,14 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                             addon.folders = folders;
                         }
 
+                        // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some(row) = ajour
-                                .catalog_search_state
-                                .catalog_rows
-                                .iter_mut()
-                                .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                            if let Some((_, _, status)) =
+                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                    flavor == *f && addon.repository_id() == Some(i.to_string())
+                                })
                             {
-                                match flavor {
-                                    Flavor::Retail => {
-                                        row.retail_install_status =
-                                            CatalogInstallStatus::Fingerprint
-                                    }
-                                    Flavor::Classic => {
-                                        row.classic_install_status =
-                                            CatalogInstallStatus::Fingerprint
-                                    }
-                                }
+                                *status = CatalogInstallStatus::Fingerprint;
                             }
                         }
 
@@ -805,25 +794,25 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                         ajour.state = AjourState::Error(err);
                         addon.state = AddonState::Ajour(Some("Error".to_owned()));
 
+                        // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some(row) = ajour
-                                .catalog_search_state
-                                .catalog_rows
-                                .iter_mut()
-                                .find(|r| r.addon.id == id.parse::<u32>().unwrap())
+                            if let Some((_, _, status)) =
+                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                    flavor == *f && addon.repository_id() == Some(i.to_string())
+                                })
                             {
-                                match flavor {
-                                    Flavor::Retail => {
-                                        row.retail_install_status = CatalogInstallStatus::Retry
-                                    }
-                                    Flavor::Classic => {
-                                        row.classic_install_status = CatalogInstallStatus::Retry
-                                    }
-                                }
+                                *status = CatalogInstallStatus::Retry;
+
+                                remove_catalog_addon = Some(addon.primary_folder_id.clone());
                             }
                         }
                     }
                 }
+            }
+
+            // Remove catalog installed addon from addons since it failed
+            if let Some(id) = remove_catalog_addon {
+                addons.retain(|a| a.primary_folder_id != id)
             }
         }
         Message::UpdateFingerprint((reason, flavor, id, result)) => {
@@ -834,31 +823,44 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                 result.is_err()
             );
 
-            if reason == DownloadReason::Install {
-                if let Some(row) = ajour
-                    .catalog_search_state
-                    .catalog_rows
-                    .iter_mut()
-                    .find(|r| r.addon.id == id.parse::<u32>().unwrap())
-                {
-                    match flavor {
-                        Flavor::Retail => {
-                            row.retail_install_status = CatalogInstallStatus::Completed
-                        }
-                        Flavor::Classic => {
-                            row.classic_install_status = CatalogInstallStatus::Completed
-                        }
-                    }
-                }
-            }
+            let mut remove_catalog_addon = None;
 
             let addons = ajour.addons.entry(flavor).or_default();
             if let Some(addon) = addons.iter_mut().find(|a| a.primary_folder_id == id) {
                 if result.is_ok() {
                     addon.state = AddonState::Ajour(Some("Completed".to_owned()));
+
+                    // Update catalog status for addon
+                    if reason == DownloadReason::Install {
+                        if let Some((_, _, status)) =
+                            ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                flavor == *f && addon.repository_id().unwrap() == i.to_string()
+                            })
+                        {
+                            *status = CatalogInstallStatus::Completed;
+                        }
+                    }
                 } else {
                     addon.state = AddonState::Ajour(Some("Error".to_owned()));
+
+                    // Update catalog status for addon
+                    if reason == DownloadReason::Install {
+                        if let Some((_, _, status)) =
+                            ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
+                                flavor == *f && addon.repository_id().unwrap() == i.to_string()
+                            })
+                        {
+                            *status = CatalogInstallStatus::Retry;
+
+                            remove_catalog_addon = Some(addon.primary_folder_id.clone());
+                        }
+                    }
                 }
+            }
+
+            // Remove catalog installed addon from addons since it failed
+            if let Some(id) = remove_catalog_addon {
+                addons.retain(|a| a.primary_folder_id != id)
             }
         }
         Message::NeedsUpdate(Ok(newer_version)) => {
@@ -1270,19 +1272,16 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             // Close settings if shown.
             ajour.is_showing_settings = false;
 
-            if let Some(row) = ajour
-                .catalog_search_state
-                .catalog_rows
-                .iter_mut()
-                .find(|r| r.addon.id == id)
-            {
-                match flavor {
-                    Flavor::Retail => row.retail_install_status = CatalogInstallStatus::Downloading,
-                    Flavor::Classic => {
-                        row.classic_install_status = CatalogInstallStatus::Downloading
-                    }
-                }
-            }
+            // Remove any existing status for this addon since we are going
+            // to try and download it again
+            ajour
+                .catalog_install_statuses
+                .retain(|(f, i, _)| if id == *i { flavor != *f } else { true });
+
+            // Add new status for this addon as Downloading
+            ajour
+                .catalog_install_statuses
+                .push((flavor, id, CatalogInstallStatus::Downloading));
 
             return Ok(Command::perform(
                 perform_fetch_latest_addon(source, id, flavor),
@@ -1362,20 +1361,12 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             Err(error) => {
                 log::error!("{}", error);
 
-                if let Some(row) = ajour
-                    .catalog_search_state
-                    .catalog_rows
+                if let Some((_, _, status)) = ajour
+                    .catalog_install_statuses
                     .iter_mut()
-                    .find(|r| r.addon.id == id)
+                    .find(|(f, i, _)| flavor == *f && id == *i)
                 {
-                    match flavor {
-                        Flavor::Retail => {
-                            row.retail_install_status = CatalogInstallStatus::Unavilable
-                        }
-                        Flavor::Classic => {
-                            row.classic_install_status = CatalogInstallStatus::Unavilable
-                        }
-                    }
+                    *status = CatalogInstallStatus::Unavilable;
                 }
             }
         },

--- a/src/gui/update.rs
+++ b/src/gui/update.rs
@@ -649,13 +649,12 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                     Ok(_) => {
                         // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some((_, _, status)) =
-                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                    flavor == *f && addon.repository_id() == Some(i.to_string())
-                                })
-                            {
-                                *status = CatalogInstallStatus::Unpacking;
-                            }
+                            update_catalog_install_status(
+                                &mut ajour.catalog_install_statuses,
+                                CatalogInstallStatus::Unpacking,
+                                flavor,
+                                addon.repository_id(),
+                            );
                         }
 
                         if addon.state == AddonState::Downloading {
@@ -680,15 +679,14 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
 
                         // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some((_, _, status)) =
-                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                    flavor == *f && addon.repository_id() == Some(i.to_string())
-                                })
-                            {
-                                *status = CatalogInstallStatus::Retry;
+                            update_catalog_install_status(
+                                &mut ajour.catalog_install_statuses,
+                                CatalogInstallStatus::Retry,
+                                flavor,
+                                addon.repository_id(),
+                            );
 
-                                remove_catalog_addon = Some(addon.primary_folder_id.clone());
-                            }
+                            remove_catalog_addon = Some(addon.primary_folder_id.clone());
                         }
                     }
                 }
@@ -751,13 +749,12 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
 
                         // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some((_, _, status)) =
-                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                    flavor == *f && addon.repository_id() == Some(i.to_string())
-                                })
-                            {
-                                *status = CatalogInstallStatus::Fingerprint;
-                            }
+                            update_catalog_install_status(
+                                &mut ajour.catalog_install_statuses,
+                                CatalogInstallStatus::Fingerprint,
+                                flavor,
+                                addon.repository_id(),
+                            );
                         }
 
                         addon.state = AddonState::Fingerprint;
@@ -796,15 +793,14 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
 
                         // Update catalog status for addon
                         if reason == DownloadReason::Install {
-                            if let Some((_, _, status)) =
-                                ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                    flavor == *f && addon.repository_id() == Some(i.to_string())
-                                })
-                            {
-                                *status = CatalogInstallStatus::Retry;
+                            update_catalog_install_status(
+                                &mut ajour.catalog_install_statuses,
+                                CatalogInstallStatus::Retry,
+                                flavor,
+                                addon.repository_id(),
+                            );
 
-                                remove_catalog_addon = Some(addon.primary_folder_id.clone());
-                            }
+                            remove_catalog_addon = Some(addon.primary_folder_id.clone());
                         }
                     }
                 }
@@ -832,28 +828,26 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
 
                     // Update catalog status for addon
                     if reason == DownloadReason::Install {
-                        if let Some((_, _, status)) =
-                            ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                flavor == *f && addon.repository_id() == Some(i.to_string())
-                            })
-                        {
-                            *status = CatalogInstallStatus::Completed;
-                        }
+                        update_catalog_install_status(
+                            &mut ajour.catalog_install_statuses,
+                            CatalogInstallStatus::Completed,
+                            flavor,
+                            addon.repository_id(),
+                        );
                     }
                 } else {
                     addon.state = AddonState::Ajour(Some("Error".to_owned()));
 
                     // Update catalog status for addon
                     if reason == DownloadReason::Install {
-                        if let Some((_, _, status)) =
-                            ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                flavor == *f && addon.repository_id() == Some(i.to_string())
-                            })
-                        {
-                            *status = CatalogInstallStatus::Retry;
+                        update_catalog_install_status(
+                            &mut ajour.catalog_install_statuses,
+                            CatalogInstallStatus::Retry,
+                            flavor,
+                            addon.repository_id(),
+                        );
 
-                            remove_catalog_addon = Some(addon.primary_folder_id.clone());
-                        }
+                        remove_catalog_addon = Some(addon.primary_folder_id.clone());
                     }
                 }
             }
@@ -1361,13 +1355,12 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             Err(error) => {
                 log::error!("{}", error);
 
-                if let Some((_, _, status)) = ajour
-                    .catalog_install_statuses
-                    .iter_mut()
-                    .find(|(f, i, _)| flavor == *f && id == *i)
-                {
-                    *status = CatalogInstallStatus::Unavilable;
-                }
+                update_catalog_install_status(
+                    &mut ajour.catalog_install_statuses,
+                    CatalogInstallStatus::Unavilable,
+                    flavor,
+                    Some(id.to_string()),
+                );
             }
         },
         Message::FetchedTukuiChangelog((addon, key, result)) => {
@@ -1738,4 +1731,18 @@ fn save_column_configs(ajour: &mut Ajour) {
     };
 
     let _ = ajour.config.save();
+}
+
+fn update_catalog_install_status(
+    statuses: &mut Vec<(Flavor, u32, CatalogInstallStatus)>,
+    new_status: CatalogInstallStatus,
+    flavor: Flavor,
+    repository_id: Option<String>,
+) {
+    if let Some((_, _, status)) = statuses
+        .iter_mut()
+        .find(|(f, i, _)| flavor == *f && repository_id == Some(i.to_string()))
+    {
+        *status = new_status;
+    }
 }

--- a/src/gui/update.rs
+++ b/src/gui/update.rs
@@ -1169,18 +1169,10 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
             let addons = ajour.addons.entry(flavor).or_default();
             addons.push(empty_addon);
 
-            let command = match source {
-                catalog::Source::Curse => Command::perform(
-                    curse_api::latest_addon(id, flavor),
-                    Message::CatalogInstallAddonFetched,
-                ),
-                catalog::Source::Tukui => Command::perform(
-                    tukui_api::latest_addon(id, flavor),
-                    Message::CatalogInstallAddonFetched,
-                ),
-            };
-
-            return Ok(command);
+            return Ok(Command::perform(
+                perform_fetch_latest_addon(source, id, flavor),
+                Message::CatalogInstallAddonFetched,
+            ));
         }
         Message::Interaction(Interaction::CatalogCategorySelected(category)) => {
             log::debug!("Interaction::CatalogCategorySelected({})", &category);
@@ -1222,42 +1214,52 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
 
             query_and_sort_catalog(ajour);
         }
-        Message::CatalogInstallAddonFetched(Ok((id, flavor, mut addon))) => {
-            log::debug!(
-                "Message::CatalogInstallAddonFetched({:?}, {:?})",
-                flavor,
-                &id
-            );
+        Message::CatalogInstallAddonFetched((flavor, id, result)) => match result {
+            Ok(mut addon) => {
+                log::debug!(
+                    "Message::CatalogInstallAddonFetched({:?}, {:?})",
+                    flavor,
+                    &id
+                );
 
-            if let Some(addons) = ajour.addons.get_mut(&flavor) {
-                // Remove the empty addon and add in our rich addon
-                if addons
-                    .iter_mut()
-                    .any(|a| a.primary_folder_id == id.to_string())
-                {
+                if let Some(addons) = ajour.addons.get_mut(&flavor) {
+                    // Remove the empty addon and add in our rich addon
+                    if addons
+                        .iter_mut()
+                        .any(|a| a.primary_folder_id == id.to_string())
+                    {
+                        addons.retain(|a| a.primary_folder_id != id.to_string());
+                    }
+
+                    addon.state = AddonState::Downloading;
+                    addons.push(addon.clone());
+
+                    let to_directory = ajour
+                        .config
+                        .get_download_directory_for_flavor(flavor)
+                        .expect("Expected a valid path");
+
+                    return Ok(Command::perform(
+                        perform_download_addon(
+                            DownloadReason::Install,
+                            ajour.shared_client.clone(),
+                            flavor,
+                            addon,
+                            to_directory,
+                        ),
+                        Message::DownloadedAddon,
+                    ));
+                }
+            }
+            Err(error) => {
+                log::error!("{}", error);
+
+                if let Some(addons) = ajour.addons.get_mut(&flavor) {
+                    // Remove the empty addon since it failed
                     addons.retain(|a| a.primary_folder_id != id.to_string());
                 }
-
-                addon.state = AddonState::Downloading;
-                addons.push(addon.clone());
-
-                let to_directory = ajour
-                    .config
-                    .get_download_directory_for_flavor(flavor)
-                    .expect("Expected a valid path");
-
-                return Ok(Command::perform(
-                    perform_download_addon(
-                        DownloadReason::Install,
-                        ajour.shared_client.clone(),
-                        flavor,
-                        addon,
-                        to_directory,
-                    ),
-                    Message::DownloadedAddon,
-                ));
             }
-        }
+        },
         Message::FetchedTukuiChangelog((addon, key, result)) => {
             log::debug!(
                 "Message::FetchedTukuiChangelog(error: {})",
@@ -1297,7 +1299,6 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
         | Message::Parse(Err(error))
         | Message::NeedsUpdate(Err(error))
         | Message::CatalogDownloaded(Err(error))
-        | Message::CatalogInstallAddonFetched(Err(error))
         | Message::DownloadedAddon((_, _, _, Err(error))) => {
             log::error!("{}", error);
 
@@ -1406,6 +1407,20 @@ async fn perform_unpack_addon(
         addon.primary_folder_id.clone(),
         install_addon(&addon, &from_directory, &to_directory).await,
     )
+}
+
+/// Unzips `Addon` at given `from_directory` and moves it `to_directory`.
+async fn perform_fetch_latest_addon(
+    source: catalog::Source,
+    source_id: u32,
+    flavor: Flavor,
+) -> (Flavor, u32, Result<Addon>) {
+    let result = match source {
+        catalog::Source::Curse => curse_api::latest_addon(source_id, flavor).await,
+        catalog::Source::Tukui => tukui_api::latest_addon(source_id, flavor).await,
+    };
+
+    (flavor, source_id, result)
 }
 
 fn sort_addons(addons: &mut [Addon], sort_direction: SortDirection, column_key: ColumnKey) {

--- a/src/gui/update.rs
+++ b/src/gui/update.rs
@@ -834,7 +834,7 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                     if reason == DownloadReason::Install {
                         if let Some((_, _, status)) =
                             ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                flavor == *f && addon.repository_id().unwrap() == i.to_string()
+                                flavor == *f && addon.repository_id() == Some(i.to_string())
                             })
                         {
                             *status = CatalogInstallStatus::Completed;
@@ -847,7 +847,7 @@ pub fn handle_message(ajour: &mut Ajour, message: Message) -> Result<Command<Mes
                     if reason == DownloadReason::Install {
                         if let Some((_, _, status)) =
                             ajour.catalog_install_statuses.iter_mut().find(|(f, i, _)| {
-                                flavor == *f && addon.repository_id().unwrap() == i.to_string()
+                                flavor == *f && addon.repository_id() == Some(i.to_string())
                             })
                         {
                             *status = CatalogInstallStatus::Retry;

